vochain/vochaininfo/vochaininfo.go

```diff
@@ -156,99 +164,100 @@ func (vi *VochainInfo) NPeers() int {
-			if pheight > 0 {
-				sync = true
-				vm++
-				n1++
-				n10++
-				n60++
-				n360++
-				n1440++
-				h1 += (height - pheight)
-				h10 += (height - pheight)
-				h60 += (height - pheight)
-				h360 += (height - pheight)
-				h1440 += (height - pheight)
-
-				if sleepSecs*n1 >= 60 && h1 > 0 {
-					a1 = int32((n1 * sleepSecs * 1000) / h1)
-					n1 = 0
-					h1 = 0
-				}
-				if sleepSecs*n10 >= 600 && h10 > 0 {
-					a10 = int32((n10 * sleepSecs * 1000) / h10)
-					n10 = 0
-					h10 = 0
-				}
-				if sleepSecs*n60 >= 3600 && h60 > 0 {
-					a60 = int32((n60 * sleepSecs * 1000) / h60)
-					n60 = 0
-					h60 = 0
-				}
-				if sleepSecs*n360 >= 21600 && h360 > 0 {
-					a360 = int32((n360 * sleepSecs * 1000) / h360)
-					n360 = 0
-					h360 = 0
-				}
-				if sleepSecs*n1440 >= 86400 && h1440 > 0 {
-					a1440 = int32((n1440 * sleepSecs * 1000) / h1440)
-					n1440 = 0
-					h1440 = 0
-				}
-			} else {
-				sync = false
-			}
-			pheight = height
-
-			vi.lock.Lock()
-			vi.height = height
-			vi.sync = sync
-			vi.avg1 = a1
-			vi.avg10 = a10
-			vi.avg60 = a60
-			vi.avg360 = a360
-			vi.avg1440 = a1440
-			var err error
-			vi.processTreeSize, err = vi.vnode.State.CountProcesses(true)
-			if err != nil {
-				log.Errorf("cannot count processes: %s", err)
-			}
-			vi.voteTreeSize, err = vi.vnode.State.VoteCount(true)
-			if err != nil {
-				log.Errorf("cannot access vote count: %s", err)
-			}
-			if sleepSecs*vm >= 60 {
-				vi.votesPerMinute = int(vi.voteTreeSize) - int(oldVoteTreeSize)
-				oldVoteTreeSize = vi.voteTreeSize
-				vm = 0
-			}
-			vi.voteCacheSize = vi.vnode.State.CacheSize()
-			vi.mempoolSize = vi.vnode.MempoolSize()
-			vi.lock.Unlock()
+		case <-t.C:
+			continue
+		case <-vi.close:
+			return
+		}
+	}
+}
 
+// trackAvgBlockTimes continously calculates the average block time
+// during the specified interval and updates vi.avgBlockTimes[interval].
+//
+// Safe for concurrent use
+func (vi *VochainInfo) trackAvgBlockTimes(interval time.Duration) {
+	// ticker t runs at shorter intervals for better sampling
+	// so (interval / 3) means the avgBlockTime[1m] is calculated/updated every 20 seconds
+	// and the avgBlockTime[24h] is updated every 8 hours. This spawns 3x goroutines tho.
+	t := time.NewTicker(interval / 3)
+	defer t.Stop()
+	for {
+		go vi.updateAvgBlockTimes(interval)
+		select {
+		case <-t.C:
+			continue
 		case <-vi.close:
 			return
 		}
 	}
 }
 
+// updateAvgBlockTimes sleeps for interval time and then updates vi.avgBlockTimes[interval].
+// (i.e. call it as a goroutine)
+func (vi *VochainInfo) updateAvgBlockTimes(interval time.Duration) {
+	// during fastsync, stats are totally skewed: don't even try.
+	if !vi.Sync() {
+		return
+	}
+	oldHeight := vi.vnode.Height()
+	select {
+	case <-time.After(interval):
+		vi.lock.Lock()
+		vi.avgBlockTimes[interval] = time.Duration(int64(interval) / int64(vi.vnode.Height()-oldHeight))
```

Comment:
I assume you're fine with integer division, because the interval is a number of nanoseconds, so it will be large enough that integer divisions will be precise enough.

You should use the Nanoseconds method, which makes that clearer: https://cs.opensource.google/go/go/+/refs/tags/go1.17.7:src/time/time.go;l=736